cni/pkg/ambient/net.go

```diff
@@ -197,6 +202,77 @@ func buildRouteFromPod(pod *corev1.Pod, ip string) ([]string, error) {
 	}, nil
 }
 
+func buildEbpfArgsByIP(ip string, isZtunnel, isRemove bool) (*ebpf.RedirectArgs, error) {
+	ipAddr, err := netip.ParseAddr(ip)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse ip(%s): %v", ip, err)
+	}
+	veth, err := getVethWithDestinationOf(ip)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get device: %v", err)
+	}
+	peerIndex, err := getPeerIndex(veth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get veth peerIndex: %v", err)
+	}
+
+	peerNs, err := getNsNameFromNsID(veth.Attrs().NetNsID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ns name: %v", err)
+	}
+
+	mac, err := getMacFromNsIdx(peerNs, peerIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ebpf.RedirectArgs{
+		IPAddrs:   []netip.Addr{ipAddr},
+		MacAddr:   mac,
+		Ifindex:   veth.Attrs().Index,
+		PeerIndex: peerIndex,
+		PeerNs:    peerNs,
+		IsZtunnel: isZtunnel,
+		Remove:    isRemove,
+	}, nil
+}
+
+// GetHostIPByRoute get the automatically chosen host ip to the Pod's CIDR
+func GetHostIPByRoute(kubeClient kubernetes.Interface) (string, error) {
+	// We assume per node POD's CIDR is the same block, so the route to the POD
+	// from host should be "same". Otherwise, there may multiple host IPs will be
+	// used as source to dial to PODs.
+	pods, err := kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(
```

Comment:
Making unbounded direct List calls to the API server is not scalable. We already have an informer, please use `s.podLister` instead

Reply:
Sure, will use shared Lister instead.